skbio/metadata/_interval.py

```diff
@@ -0,0 +1,699 @@
+        # this ``Interval`` object will be returned
+        # multiple times. So we need to remove duplicates.
+        seen = set()
+        for intvl in intvls:
+            if id(intvl) not in seen:
+                seen.add(id(intvl))
+                yield intvl
+
+    def _query_attribute(self, metadata, intervals=None):
+        """Yield ``Interval`` objects based on query attributes.
+
+        Parameters
+        ----------
+        metadata : dict or ``None``
+            If it is ``None``, return empty iterator; if it is
+            ``{}``, return an interator of all the ``Interval``
+            objects.
+        intervals : an iterable of ``Interval`` objects
+        """
+        if metadata is None:
+            return
+            yield
+
+        if intervals is None:
+            intervals = self._intervals
+
+        for intvl in intervals:
+            for (key, value) in metadata.items():
+                if (key not in intvl.metadata or
+                        intvl.metadata[key] != value):
+                    break
+            else:
+                yield intvl
+
+    @experimental(as_of='0.5.0-dev')
+    @_rebuild_tree
+    def query(self, locations=None, metadata=None):
+        """Yield ``Interval`` object with the locations and attributes.
+
+        The ``Interval`` objects must meet both requirements: 1) overlap
+        with any of the spans specified by ``locations``; 2) satisfy
+        ``metadata`` specification. For instance, you can identify
+        all the recA genes that overlap with (10, 100) or (900, 1000)
+        with this code ``interval_metadata.query([(10, 100),
+        (900, 1000)], {'gene': 'recA'})``.
+
+        Parameters
+        ----------
+        locations : iterable of tuples of int pair
+            Specifies locations to look for the ``Interval``
+            objects. An satisfying interval feature only need to overlap with
+            one location. Default (``None``) means all ``Interval``s meet
+            this requirement.
+
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            ``Interval`` object. It specifies what metadata keywords and
+            values to look for. Default (``None``) means all ``Interval``s
+            meet this requirement.
+
+        Returns
+        -------
+        generator, Interval
+            A generator of Interval objects satisfying the search criteria.
+        """
+        if locations is None:
+            for intvl in self._query_attribute(metadata):
+                yield intvl
+        else:
+            for loc in locations:
+                intvls = self._query_interval(loc)
+                if metadata is None:
+                    metadata = {}
+                for intvl in self._query_attribute(metadata, intvls):
+                    yield intvl
+
+    @experimental(as_of='0.5.0-dev')
+    def drop(self, locations=None, metadata=None):
```

Comment:
ok. I will change the interface to accept an iterable of `Interval` instead of a single `Interval`. Reasons:

1. For single interval, you can call `Interval.drop()` instead.

2. only loop thru the interval list once to drop all the given intervals.

do you guys agree.